File: test_regress/t/t_trace_vcd_comment.cpp
// -*- mode: C++; c-file-style: "cc-mode" -*-
//
// DESCRIPTION: Verilator: Verilog Test module
//
// This file ONLY is placed under the Creative Commons Public Domain.
// SPDX-FileCopyrightText: 2026 Wilson Snyder
// SPDX-License-Identifier: CC0-1.0

#include <verilated.h>
#include <verilated_vcd_c.h>

#include <memory>

#include VM_PREFIX_INCLUDE

int main(int argc, char** argv) {
    Verilated::debug(0);
    Verilated::traceEverOn(true);
    Verilated::commandArgs(argc, argv);

    const std::unique_ptr<VM_PREFIX> top{new VM_PREFIX{"top"}};

    const std::unique_ptr<VerilatedVcdC> tfp{new VerilatedVcdC};
    top->trace(tfp.get(), 99);
    tfp->open(VL_STRINGIFY(TEST_OBJ_DIR) "/simcomment.vcd");

    tfp->comment("comment before first dump");

    top->clk = 0;
    for (uint64_t time = 0; time < 10; ++time) {
        top->clk = !top->clk;
        top->eval();
        tfp->dump(time);
        if (time == 3) {
            tfp->comment("single line comment at t=3");
        } else if (time == 5) {
            tfp->comment("multi line comment at t=5\nsecond line");
        }
    }

    tfp->close();
    top->final();
    printf("*-* All Finished *-*\n");
    return 0;
}

File: test_regress/t/t_trace_vcd_comment.py
#!/usr/bin/env python3
# DESCRIPTION: Verilator: Verilog Test driver/expect definition
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of either the GNU Lesser General Public License Version 3
# or the Perl Artistic License Version 2.0.
# SPDX-FileCopyrightText: 2026 Wilson Snyder
# SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0

import vltest_bootstrap

test.scenarios('vlt_all')
test.top_filename = "t_trace_cat.v"

test.compile(make_top_shell=False,
             make_main=False,
             v_flags2=["--trace-vcd --exe", test.pli_filename])

test.execute()

vcd = test.obj_dir + "/simcomment.vcd"

test.file_grep(vcd, r'\$comment comment before first dump \$end\n')
test.file_grep(vcd, r'\$comment single line comment at t=3 \$end\n')
test.file_grep(vcd, r'\$comment multi line comment at t=5\nsecond line \$end\n')

test.passes()
